hashintel/hash
# Make 0.3 blocks load in HASH

## 🌟 What is the purpose of this PR?

Updates HASH to load blocks compliant with the `0.3` specification.

This means that we can reintroduce blocks that are currently not working in HASH by updating them to be `0.3` compliant. Prior to this PR, HASH had been partially updated to `0.3` in a way which meant no blocks were working (except text blocks, which do not rely on any message passing under the graph service).

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1203543021352032/1203714761594036/f) _(internal)_

## 🔍 What does this change?

- Installs new canary versions of packages in `hash/frontend`
- Removes redundant code related to:
  - an earlier data mapping prototype which is not worth retaining, as it depends heavily on a previous way of linking entities. this includes code to fetch a block's schema, which the data mapping prototype used
  - rewriting entities to be in the shape defined in the BP. In `0.3`, HASH and BP entity shape will be consistent
- updates `block-loader.tsx` to fetch the subgraph rooted at the block entity. There are a few temporary measures involved which are covered in 'known issues'
- replaces the `updateEntity` shim with the new version imported from `@blockprotocol/graph@canary`
- **driveby**: fix filename for consistently, `loadremote-block.tsx` -> `load-remote-block.tsx`
- **driveby**: update the branded type to use `{ [_ in Kind]: true }>` instead of `{ [_ in Kind]: void }>` – makes a bit more sense semantically and didn't seem to be any objection to doing this

## ⚠️ Known issues

- There is a mismatch between the type of `Subgraph` in `@blockprotocol/graph` and in `hash/subgraph` – particularly in the identifiers for entities (`version` versus `versionId`). This leads to some temporary assertions in the code. 
- Because blocks use the `graph` package and expect types in its shape, there is some _adding_ fields they expect (`versionId`) so that they can correctly look up items in the subgraph map
- Each block fetching its own `subgraph` is a temporary measure that should be replaced by fetching the subgraph rooted at the `Page` to a sufficient depth to also fetch the block's linked entities. Currently, the special-cased `getPage` query terminates the query at the block entity. This should be fixed by removing the special casing of page retrieval in favour of using the general `getEntity` to fetch the appropriate subgraph.
- Paragraph blocks are automatically loaded on every page, and require temporarily adding properties in the pre-`0.3` shape to stop them crashing.

## 🐾 Next steps

- Sort out mismatch between `@blockprotocol/graph` and types in `hash`.
- Update more blocks to `0.3`

## ❓ How to test this?



1. `yarn install`
2.  in `blocks/code` run `yarn dev` so that the block's dev server is running
3. run the HASH app by `yarn dev`
4. Navigate to a page, click the context menu next to a block, 'load block', and paste in the code block's URL (should be `http://localhost:9090`
